src/Common/src/TypeSystem/Common/TypeSystemConstaintsHelpers.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+using System.Diagnostics;
+
+namespace Internal.TypeSystem
+{
+    public static class TypeSystemConstraintsHelpers
+    {
+        private static bool VerifyGenericParamConstraint(Instantiation typeInstantiation, Instantiation methodInstantiation, GenericParameterDesc genericParam, TypeDesc instantiationParam)
+        {
+            // Check class constraint
+            if (genericParam.Constraints.HasFlag(GenericConstraints.ReferenceTypeConstraint) && !instantiationParam.IsGCPointer)
+                return false;
+
+            // Check default constructor constraint
+            if (genericParam.Constraints.HasFlag(GenericConstraints.DefaultConstructorConstraint))
+            {
+                if (!instantiationParam.IsDefType)
+                    return false;
+
+                if (!instantiationParam.IsValueType && instantiationParam.GetDefaultConstructor() == null)
+                    return false;
+            }
+
+            // Check struct constraint
+            if (genericParam.Constraints.HasFlag(GenericConstraints.NotNullableValueTypeConstraint))
+            {
+                if (!instantiationParam.IsDefType)
```

Comment:
Nit: the `!instantiationParam.IsValueType` check will cover this. We don't need the two lines.